resolver-dns/src/main/java/io/netty/resolver/dns/DefaultDnsCache.java

```diff
@@ -289,4 +252,99 @@ public String toString() {
             }
         }
     }
+
+    // Directly extend AtomicReference for intrinsics and also to keep memory overhead low.
+    private static final class Entries extends AtomicReference<List<DefaultDnsCacheEntry>> {
+
+        Entries(DefaultDnsCacheEntry entry) {
+            super(Collections.singletonList(entry));
+        }
+
+        void add(DefaultDnsCacheEntry e) {
+            if (e.cause() == null) {
+                for (;;) {
+                    List<DefaultDnsCacheEntry> entries = get();
+                    if (!entries.isEmpty()) {
+                        final DefaultDnsCacheEntry firstEntry = entries.get(0);
+                        if (firstEntry.cause() != null) {
+                            assert entries.size() == 1;
+                            if (compareAndSet(entries, Collections.singletonList(e))) {
+                                firstEntry.cancelExpiration();
+                                return;
+                            } else {
+                                // Need to try again as CAS failed
+                                continue;
+                            }
+                        }
+                        // Create a new List for COW semantics
+                        List<DefaultDnsCacheEntry> newEntries = new ArrayList<DefaultDnsCacheEntry>(entries.size() + 1);
+                        newEntries.addAll(entries);
+                        newEntries.add(e);
+                        if (compareAndSet(entries, newEntries)) {
+                            return;
+                        }
+                    } else if (compareAndSet(entries, Collections.singletonList(e))) {
+                        return;
+                    }
+                }
+            } else {
+                List<DefaultDnsCacheEntry> entries = getAndSet(Collections.singletonList(e));
+                cancelExpiration(entries);
+            }
+        }
+
+        boolean remove(DefaultDnsCacheEntry entry) {
+            for (;;) {
+                List<DefaultDnsCacheEntry> entries = get();
+                int size = entries.size();
+                if (size == 0) {
+                    // If the list is empty we just return early and so not allocate a new ArrayList.
+                    if (compareAndSet(entries, Collections.<DefaultDnsCacheEntry>emptyList())) {
+                        return false;
+                    }
+                } else if (size == 1) {
+                    if (entries.get(0).equals(entry)) {
+                        // If the list is empty we just return early and so not allocate a new ArrayList.
+                        if (compareAndSet(entries, Collections.<DefaultDnsCacheEntry>emptyList())) {
+                            return false;
+                        }
```

Comment:
it looks like we should return `true` in an else condition? consider simplifying the code the control flow is easier to follow:

```java
return entries.get(0).equals(entry) && compareAndSet(entries, Collections.<DefaultDnsCacheEntry>emptyList());
```

Reply:
doh! Yeah... let me do what you suggested.

Reply:
Actually no... there was something else wrong.

This is correct:

```java
                    if (entries.get(0).equals(entry)) {
                        // If the list is empty we just return early and so not allocate a new ArrayList.
                        if (compareAndSet(entries, Collections.<DefaultDnsCacheEntry>emptyList())) {
                            return false;
                        } else {
                            // try again as the CAS failed
                            continue;
                        }
                    }
                    return false;
```

@Scottmitch WDYT ?

Reply:
sorry this was not correct as well. Pushed a new version which should be correct